pkg/storage/chunk/azure/blob_storage_client.go

```diff
@@ -220,205 +136,142 @@ func (b *BlobStorage) GetObject(ctx context.Context, objectKey string) (io.ReadC
+		return azblob.ServiceClient{}, nil
 	}
 
-	tc := azblob.NewTokenCredential(spt.Token().AccessToken, func(tc azblob.TokenCredential) time.Duration {
-		err := spt.Refresh()
-		if err != nil {
-			// something went wrong, prevent the refresher from being triggered again
-			return 0
-		}
-
-		// set the new token value
-		tc.SetToken(spt.Token().AccessToken)
+	return azblob.NewServiceClient(fmt.Sprintf(accountURL(cfg.Environment), cfg.AccountName), cred, opts)
+}
 
-		// get the next token slightly before the current one expires
-		return time.Until(spt.Token().Expires()) - 10*time.Second
-	})
+func sharedKeyService(cfg *BlobStorageConfig, opts *azblob.ClientOptions) (azblob.ServiceClient, error) {
+	cred, err := azblob.NewSharedKeyCredential(cfg.AccountName, cfg.AccountKey.String())
+	if err != nil {
+		return azblob.ServiceClient{}, err
+	}
 
-	return &tc, nil
+	return azblob.NewServiceClientWithSharedKey(fmt.Sprintf(accountURL(cfg.Environment), cfg.AccountName), cred, opts)
 }
 
-func (b *BlobStorage) fetchMSIToken() (*adal.ServicePrincipalToken, error) {
-	// msiEndpoint is the well known endpoint for getting MSI authentications tokens
-	// msiEndpoint := "http://169.254.169.254/metadata/identity/oauth2/token" for production Jobs
-	msiEndpoint, _ := adal.GetMSIVMEndpoint()
+func accountURL(env string) string {
+	return endpoints[env].blobURLFmt
+}
 
-	// both can be empty, systemAssignedMSI scenario
-	spt, err := adal.NewServicePrincipalTokenFromMSI(msiEndpoint, "https://storage.azure.com/")
-	if err != nil {
-		return nil, err
+func getHTTPClient(hedgingCfg hedging.Config, hedging bool) (*http.Client, error) {
+	client := defaultClientFactory()
+	if hedging {
+		return hedgingCfg.ClientWithRegisterer(client, prometheus.WrapRegistererWithPrefix("loki_", prometheus.DefaultRegisterer))
 	}
-
-	return spt, spt.Refresh()
+	return client, nil
 }
 
 // List implements chunk.ObjectClient.
 func (b *BlobStorage) List(ctx context.Context, prefix, delimiter string) ([]chunk.StorageObject, []chunk.StorageCommonPrefix, error) {
 	var storageObjects []chunk.StorageObject
 	var commonPrefixes []chunk.StorageCommonPrefix
 
-	for marker := (azblob.Marker{}); marker.NotDone(); {
-		if ctx.Err() != nil {
-			return nil, nil, ctx.Err()
-		}
-
-		err := instrument.CollectedRequest(ctx, "azure.List", instrument.NewHistogramCollector(b.metrics.requestDuration), instrument.ErrorCode, func(ctx context.Context) error {
-			listBlob, err := b.containerURL.ListBlobsHierarchySegment(ctx, marker, delimiter, azblob.ListBlobsSegmentOptions{Prefix: prefix})
-			if err != nil {
-				return err
-			}
-
-			marker = listBlob.NextMarker
-
-			// Process the blobs returned in this result segment (if the segment is empty, the loop body won't execute)
-			for _, blobInfo := range listBlob.Segment.BlobItems {
+	err := instrument.CollectedRequest(ctx, "azure.List", instrument.NewHistogramCollector(b.metrics.requestDuration), instrument.ErrorCode, func(ctx context.Context) error {
```

Comment:
The new API means that this instrumentation is now for how long it takes to list all the blobs rather than each page of blobs